src/expr/src/scalar/func.rs

```diff
@@ -1166,6 +1166,82 @@ fn add_interval<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError>
         .map(Datum::from)
 }
 
+fn bit_not_int16<'a>(a: Datum<'a>) -> Datum<'a> {
+    Datum::from(!a.unwrap_int16())
+}
+
+fn bit_not_int32<'a>(a: Datum<'a>) -> Datum<'a> {
+    Datum::from(!a.unwrap_int32())
+}
+
+fn bit_not_int64<'a>(a: Datum<'a>) -> Datum<'a> {
+    Datum::from(!a.unwrap_int64())
+}
+
+fn bit_and_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int16() & b.unwrap_int16())
+}
+
+fn bit_and_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() & b.unwrap_int32())
+}
+
+fn bit_and_int64<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int64() & b.unwrap_int64())
+}
+
+fn bit_or_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int16() | b.unwrap_int16())
+}
+
+fn bit_or_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() | b.unwrap_int32())
+}
+
+fn bit_or_int64<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int64() | b.unwrap_int64())
+}
+
+fn bit_xor_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int16() ^ b.unwrap_int16())
+}
+
+fn bit_xor_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() ^ b.unwrap_int32())
+}
+
+fn bit_xor_int64<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int64() ^ b.unwrap_int64())
+}
+
+fn bit_shift_left_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    // widen to i32 and then cast back to i16 in order to align with the behavior of Postgres
+    // when the number of bits is in the 16-31 range
```

Comment:
Is there a link to the pg docs about this that can be included in this comment?

Reply:
Sadly this part is based on empirical analysis on the behavior of Postgres in cases where the number of shifted bits exceeds the integer length. I've added tests in `arithmetic.slt` ([here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589) and [here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589)).

I can reference the tests file or inline a specific example.

Reply:
Hm, you can link to this section of pg source code https://github.com/postgres/postgres/blob/REL_14_STABLE/src/backend/utils/adt/int.c#L1460-L1476

Interestingly [`i16::overflowing_shl`](https://doc.rust-lang.org/std/primitive.i16.html#method.overflowing_shl) takes a u32 argument so this can also be written like so:

```rust
a.unwrap_int16().oveflowing_shl(b.unwrap_i32() as u32).0
```

Reply:
According to Postgres we should get:

```
1::int2 <<  1 = 2
1::int2 << 15 = -32768
1::int2 << 16 = 0
1::int2 << 17 = 0 
...
1::int2 << 31 = 0
1::int2 << 32 = 1
1::int2 << 33 = 1
```
and the way to get this behavior is apply the `<<` operation on a lhs that is temporarily widened from `i16` to `i32`.

Reply:
After further discussion and digging by @petrosagg, seems like an instance of [C's integer promotion rules](http://www.idryman.org/blog/2012/11/21/integer-promotion/#:~:text=In%20C99%2C%20integer%20promotion%20is,are%20called%20the%20integer%20promotions.) in action.

Put a comment on that in the `bit_shift_left_int16` and `bit_shift_right_int16` bodies.